### DISCUSSION TITLE
Ability to reference the test directory in the pom.xml under test?

### DISCUSSION THREAD

**Discussion by aegershman** [khmarbaise/maven-it-extension#193] [Q&A] [answered]
Given a `pom.xml` under test (e.g., under a directory like `src/test/resources-its/.../SomeIT/some_test/**`), how should configuration paths within the `pom.xml` reference the same directory under test?

**background info**

What I'm trying to do is have the `pom.xml` under test reference another file within the test folder. I'd like to be able to use maven dynamic property resolution (such as `${project.basedir}/rewrite.yml`) within the `pom.xml` to reference the file rather than provide full details on the path.

Here's a specific example of what I mean: https://github.com/openrewrite/rewrite-maven-plugin/blob/0b51dc8b0da87b4652f2b605c4e09e14e69afe51/src/test/resources-its/org/openrewrite/maven/RewriteDryRunIT/multi_module_project/pom.xml#L27-L29 

I'd like the `pom.xml` to be able to reference the file `rewrite.yml` which is at the same root level as the `pom.xml` in the test folder (specifically: https://github.com/openrewrite/rewrite-maven-plugin/blob/0b51dc8b0da87b4652f2b605c4e09e14e69afe51/src/test/resources-its/org/openrewrite/maven/RewriteDryRunIT/multi_module_project/rewrite.yml)

If I have the `pom.xml` reference `${project}`, it tries to use the full path to the `pom.xml`, which is close. But I need the base directory. If I use `${project.basedir}`, I get the base directory to the overall plugin, which isn't quite right either.

Currently I'm partially referencing the full path using `                            ${maven.multiModuleProjectDirectory}/src/test/resources-its/org/openrewrite/maven/RewriteDryRunIT/multi_module_project/rewrite.yml`, but I'd like to be able to do something like `${maven.multiModuleProjectDirectory.basedir}/rewrite.yml` (or something to that effect) so I don't need to qualify each test with the path.

Is there any way to do this? Thanks in advance for any guidance, I appreciate the time.

**Comment by khmarbaise** [comment, marked as answer]
So I started to investigate the problem.

The problem with `maven.multiModuleProjectDirectory` is that it is interpolated of the `mvn.cmd` or `mvn` script and could be wrong in some situations. In exactly the situation happens within the integration tests because you have a project in the directory `maven-it/org/openwrite/maven/RewriteDryRun/multi_module_project`. The `mvn` process will be started with that directory and the `mvn` script tries to find the root which results in finding it on the level `rewrite-maven-plugin` (root level) because there `.mvn` has been found. That's the reason why you had to use `${maven.multiModuleProjectDirectory}/src/test/resources-its/org/openrewrite/maven/RewriteDryRunIT/multi_module_project/rewrite.yml`.

An alternative solution to find configurations for builds is either via classpath (usually via a dependency separate module within the reactor like  [maven-checkstyle-plugin](https://maven.apache.org/plugins/maven-checkstyle-plugin/examples/multi-module-config.html) is doing it or [versions-maven-plugin](https://www.mojohaus.org/versions-maven-plugin/version-rules.html) or via external resource URI ). 

My recommendation is not to use `maven.multiModuleProjectDirectory` because it's an internal property and should not being used within a plugin or build configurations. Apart from being removed with later versions of Maven.

I've taken a look into the code why it's needed. To find the `baseDir` this can be easily done by using:
```java
    @Parameter(defaultValue = "${session}", readonly = true, required = true)
    protected MavenSession session;

   ....
        MavenProject mavenProject = session.getProjectDependencyGraph().getSortedProjects().get(0);
```
This would give you the mavenProject at the top level...within the reactor. 

Apart from that I will take a deeper look into that see if I could find a different (maybe better solution) for the problem. see #194 

What me currently puzzles is that `${project.basedir}` produces something weird from my perspective...

**Reply by aegershman** [reply]
Right on, okay. These are good takeaways. I'll keep an eye on #194 (though it's not an immediate personal priority, please don't feel compelled to spend inordinate time specifically on this particular use case), and I'll log something on our plugin to reconsider our usage of `${maven.multiModuleProjectDirectory}` altogether. Thanks for your advice and expertise @khmarbaise and thanks for this project altogether, it honestly makes life better 👍

**Reply by khmarbaise** [reply]
This particular use case shows me some parts which needed to be reconsidered. Every time I see such things it helps me better to understand what people really need (apart from my own requests and ideas). I have a lot of examples (done of my own etc.) but other people have different idea / use. cases etc. That's perfect. 

Thanks for your information. Feel free to ask more questions... I appreciate to help.

And great to hear that it helps. 🚀

**Reply by splatch** [reply]
Another way I found, which does not require plugin changes is this custom extension: https://github.com/pascalgn/properties-maven-extension. It provides a placeholder `${git.dir.worktree}` which will contain absolute path to project checkout, which in many cases might be base for making relative paths.
